includes/class-webfinger.php

```diff
@@ -94,4 +94,110 @@ public static function resolve( $resource ) {
 		\set_transient( $transient_key, $link, HOUR_IN_SECONDS ); // Cache the error for a shorter period.
 		return $link;
 	}
+
+	/**
+	 * Convert a URI string to an identifier and its host.
+	 * Automatically adds acct: if it's missing.
+	 *
+	 * @param string $url The URI (acct:, mailto:, http:, https:)
+	 *
+	 * @return WP_Error|array Error reaction or array with
+	 *                        identifier and host as values
+	 */
+	public static function get_identifier_and_host( $url ) {
+		// remove leading @
+		$url = ltrim( $url, '@' );
+
+		if ( ! preg_match( '/^([a-zA-Z+]+):/', $url, $match ) ) {
+			$identifier = 'acct:' . $url;
+			$scheme = 'acct';
+		} else {
+			$identifier = $url;
+			$scheme = $match[1];
+		}
+
+		$host = null;
+
+		switch ( $scheme ) {
+			case 'acct':
+			case 'mailto':
+			case 'xmpp':
+				if ( strpos( $identifier, '@' ) !== false ) {
+					$host = substr( $identifier, strpos( $identifier, '@' ) + 1 );
+				}
+				break;
+			default:
+				$host = wp_parse_url( $identifier, PHP_URL_HOST );
+				break;
+		}
+
+		if ( empty( $host ) ) {
+			return new WP_Error( 'invalid_identifier', __( 'Invalid Identifier', 'activitypub' ) );
+		}
+
+		return array( $identifier, $host );
+	}
+
+	/**
+	 * Get the WebFinger data for a given URI
+	 *
+	 * @param string $identifier The Identifier: <identifier>@<host>
+	 * @param string $host       The Host: <identifier>@<host>
+	 *
+	 * @return WP_Error|array Error reaction or array with
+	 *                        identifier and host as values
+	 */
+	public static function get_data( $identifier, $host ) {
+		$webfinger_url = 'https://' . $host . '/.well-known/webfinger?resource=' . rawurlencode( $identifier );
+
+		$response = wp_remote_get(
```

Comment:
Especially because this is pinging remote servers based on user-provided information, let's use `wp_safe_remote_get` here